config/config_conditions.go

```diff
@@ -0,0 +1,176 @@
+		result += fmt.Sprintf("Ignore if %s", c.rules(c.Denies))
+	}
+	if len(c.Allows) > 0 {
+		result += fmt.Sprintf("%sRun if %s", sep, c.rules(c.Allows))
+	}
+	return result
+}
+
+// ShouldRun returns whether or not the current project should be run based on its run conditions and the variables in its options.
+func (c RunConditions) ShouldRun() bool {
+	answer := true
+	for _, deny := range c.Denies {
+		ok, err := deny.isTrue()
+		if err != nil {
+			deny.options().Logger.Warningf("Ignoring project because %v in %s", err, deny)
+			answer = false
+		}
+		if ok {
+			deny.options().Logger.Warningf("Ignoring project because of ignore rule %s", deny)
+			answer = false
+		}
+	}
+	if !answer {
+		// There is at least one deny condition
+		return false
+	}
+
+	if len(c.Allows) == 0 {
+		// There is no run_if condition, so we
+		return true
+	}
+
+	options, hasErr := c.Allows[0].options(), false
+	answer = false
+	for _, allow := range c.Allows {
+		if result, err := allow.isTrue(); err != nil {
+			allow.options().Logger.Warningf("Ignoring project because %v in %s", err, allow)
+			hasErr = true
+		} else if result {
+			answer = true
+		}
+	}
+
+	if answer && !hasErr {
+		return true
+	}
+
+	if !answer {
+		options.Logger.Warningf("Ignoring project because running condition is not met: %s", c.rules(c.Allows))
+	}
+	return false
+}
+
+// Merge combines RunConditions from another source into the current one
+func (c *RunConditions) Merge(imported RunConditions) {
+	c.Allows = append(c.Allows, imported.Allows...)
+	c.Denies = append(c.Denies, imported.Denies...)
+}
+
+func (c RunConditions) rules(conditions []Condition) string {
+	result := make([]string, len(conditions))
+	for i, condition := range conditions {
+		result[i] = fmt.Sprint(condition)
+	}
+	return strings.Join(result, " or ")
+}
+
+func (c *RunConditions) init(options *options.TerragruntOptions) (err error) {
+	// There is a problem with the native hcl parser and we lost grouping in conditions
+	// so, we reinitialize the structure with the raw information
+	defer func() { err = errors.Trap(err, recover()) }()
+	dict := func(i interface{}) map[string]interface{} { return collections.AsDictionary(i).AsMap() }
+	list := func(i interface{}) []interface{} { return collections.AsList(i).AsArray() }
+	conditions := list(options.TerragruntRawConfig.Get("run_conditions"))
+	c.Allows, c.Denies = nil, nil
+	for _, condition := range conditions {
+		condition := dict(condition)
+		for _, allow := range list(condition["run_if"]) {
+			c.Allows = append(c.Allows, Condition(dict(allow)).init(options))
+		}
+		for _, deny := range list(condition["ignore_if"]) {
+			c.Denies = append(c.Denies, Condition(dict(deny)).init(options))
+		}
+	}
+	return
+}
+
+// Condition defines a single condition
+type Condition map[string]interface{}
+
+func (c Condition) options() *options.TerragruntOptions {
+	return c[optionsConfigKey].(*options.TerragruntOptions)
+}
+
+func (c Condition) getVariable(v string) string {
```

Comment:
I guess having a reference to the TerragruntOptions in Condition is not that bad. But I think that method belongs in the TerragruntOptions class or elsewhere.

Reply:
Yeah, it may be a good idea to make it part of the TerragruntOptions class.